### PR TITLE
Support for server ACLs

### LINKED ISSUE #1259: Implement Server ACLs
https://matrix.org/docs/spec/server_server/latest#server-access-control-lists-acls

We need to do this before people start using this properly, else it will cause split-brained rooms and will result in dendrite instances being banned.

### PR DESCRIPTION
**Pull request by neilalexander**
This PR adds support for server ACLs.

Closes #1259.

### CODE REVIEW

File: currentstateserver/consumers/roomserver.go [hunk 1]
```diff
@@ -76,6 +79,10 @@ func (c *OutputRoomEventConsumer) onNewRoomEvent(
 ) error {
 	ev := msg.Event
 
+	if ev.Type() == "m.room.server_acl" {
```
Comment (Kegsay) [inline currentstateserver/consumers/roomserver.go:82]: Check the state key as well please, it should exist.

File: currentstateserver/inthttp/client.go [hunk 2]
```diff
@@ -108,3 +109,13 @@ func (h *httpCurrentStateInternalAPI) QueryKnownUsers(
 	apiURL := h.apiURL + QueryKnownUsersPath
 	return httputil.PostJSON(ctx, span, h.httpClient, apiURL, req, res)
 }
+
+func (h *httpCurrentStateInternalAPI) QueryServerBannedFromRoom(
+	ctx context.Context, req *api.QueryServerBannedFromRoomRequest, res *api.QueryServerBannedFromRoomResponse,
+) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "QueryKnownUsers")
```
Comment (Kegsay) [inline currentstateserver/inthttp/client.go:116]:
```suggestion
	span, ctx := opentracing.StartSpanFromContext(ctx, "QueryServerBannedFromRoom")
```

File: currentstateserver/inthttp/server.go [hunk 3]
```diff
@@ -90,4 +90,17 @@ func AddRoutes(internalAPIMux *mux.Router, intAPI api.CurrentStateInternalAPI) {
 			return util.JSONResponse{Code: http.StatusOK, JSON: &response}
 		}),
 	)
+	internalAPIMux.Handle(QuerySharedUsersPath,
```
Comment (Kegsay) [inline currentstateserver/inthttp/server.go:93]:
```suggestion
	internalAPIMux.Handle(QueryServerBannedFromRoomPath,
```

File: currentstateserver/acls/acls.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	// Check if the hostname matches one of the allowed regexes. If it does then
+	// the server is NOT banned from the room.
+	for _, expr := range acls.allowedRegexes {
+		if expr.MatchString(string(serverName)) {
+			return false
+		}
+	}
+	// If we've got to this point then we haven't matched any regexes or an IP
+	// hostname if disallowed. The spec calls for default-deny here.
+	return true
```
Comment (Kegsay) [inline currentstateserver/acls/acls.go:141]: Please add unit tests for this function as it's critical the logic here is correct.

File: currentstateserver/acls/acls_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	if acls.IsServerBannedFromRoom("1.2.3.4", roomID) {
+		t.Fatal("Expected 1.2.3.4 to be allowed but wasn't")
+	}
+	if acls.IsServerBannedFromRoom("1.2.3.4:2345", roomID) {
+		t.Fatal("Expected 1.2.3.4 to be allowed but wasn't")
+	}
+	if !acls.IsServerBannedFromRoom("foo.com", roomID) {
+		t.Fatal("Expected foo.com to be banned but wasn't")
```
Comment (Kegsay) [inline currentstateserver/acls/acls_test.go:52]: How do we handle subdomains e.g `sub.foo.com`?

Comment (neilalexander) [reply currentstateserver/acls/acls_test.go:52]: The spec makes no special recognition for subdomains - they would need to be wildcarded.

**Pull request merged**